Added signal averager using decorators. These are separate files rath…

andrewvanderbyl opened the pull request:
…er than an overwrite so that a comparison can be made easily. The intention is not to merge into one file but keep as separate examples.

andrewvanderbyl committed:
Added signal averager using decorators. These are separate files rather than an overwrite so that a comparison can be made easily. The intention is not to merge into one file but keep as separate examples.

james-smith-za commented on SigAverager/SigAverager/signal_averager_decorator.py:
```diff
@@ -0,0 +1,89 @@
+
+    def inner(f):
+        def wrapper(*args, **kwargs):
+            rv = np.zeros(args[4])
```
So... an implicit assumption behind decorators (which I thought was obvious but clearly I was mistaken!) is that they shouldn't need (or in fact have) any knowledge about the function to which they'll be applied.

james-smith-za commented on SigAverager/SigAverager/signal_averager_decorator.py:
```diff
@@ -0,0 +1,89 @@
+            rv = np.zeros(args[4])
+            for i in range(n):
+                print(f"Running {f.__name__}:{i}")
+                rv = f(rv, *args, **kwargs)
```
I get the following output from pytest:

```
        # Compute Averaged Signal
>       averaged_signal = SigAverager.signal_averager(
            cw_scale, cw_freq, sampling_frequency, noise_scale, num_samples, n_iter
        )

testing/test_dsp_signal_averager.py:45:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (0.01, 75000000.0, 1712000000.0, 0.4, 8192, 2048), kwargs = {}, rv = array([0., 0., 0., ..., 0., 0., 0.]), i = 0

    def wrapper(*args, **kwargs):
        rv = np.zeros(args[4])
        for i in range(n):
            print(f"Running {f.__name__}:{i}")
>           rv = f(rv, *args, **kwargs)
E           TypeError: signal_averager() takes 6 positional arguments but 7 were given

../../.venv/lib/python3.8/site-packages/SigAverager/signal_averager_decorator.py:22: TypeError
```

So what's happening here is that you're adding an extra argument in the front that the wrapped function isn't expecting.

I think you could probably get it to work, if you lopped the first one off `args`, in the following manner:
```suggestion
                rv = f(rv, *args[1:], **kwargs)
```

james-smith-za commented on SigAverager/SigAverager/signal_averager_decorator.py:
```diff
@@ -0,0 +1,89 @@
+    noise_scale: float
+        Factor to scale generated noise.
+
+    Decorator:
+    Number_of_iterations: int
+        Nmber of iterations to average signal
```
Similarly, your function would not ordinarily know anything about what's decorating it.

andrewvanderbyl committed:
Added fix to SigAverager using decorators

This commit added a fix to the recursion of the return value used when
doing signal averaging.

No SARAO JIRA ticket as it's a private repo.

james-smith-za commented on .vscode/launch.json:
```diff
@@ -2,6 +2,7 @@
+    // "program": "${workspaceFolder}/SigAverager/testing/test_dsp_signal_averager_decorator.py"
     "version": "0.2.0",
     "configurations": [
         {
```
May be worth adding the following line to your `.gitignore`:
```
.vscode
```
